File: php/src/SdkClient.php
<?php

declare(strict_types=1);

namespace TwoFinance\Sdk;

final class SdkClient
{
    public readonly AuthClient $auth;
    public readonly NetworkClient $network;
    public readonly AnalyticsClient $analytics;
    public readonly OrchestratorClient $orchestrator;
    public readonly MCPClient $mcp;
    public readonly TradingControlClient $tradingControl;
    public readonly MatchEngineClient $matchEngine;
    public readonly KeyStoreClient $keystore;
    public readonly HummingbotClient $hummingbot;
    public readonly WiseClient $wise;
    public readonly AirwallexClient $airwallex;
    public readonly PlannerClient $planner;

    public function __construct(
        public readonly SdkConfig $config,
        ?callable $transport = null,
    ) {
        $tokenSource = $config->tokenSource;
        $this->auth = new AuthClient(
            new ServiceClient($config->authUrl, $transport, $tokenSource),
            $config->authRealm,
            $config->authClientId,
            $config->authPhoneClientId,
        );
        $this->network = new NetworkClient(new ServiceClient($config->networkUrl, $transport, $tokenSource));
        $this->analytics = new AnalyticsClient(new ServiceClient($config->analyticsUrl, $transport, $tokenSource));
        $this->orchestrator = new OrchestratorClient(new ServiceClient($config->orchestratorUrl, $transport, $tokenSource));
        $this->mcp = new MCPClient(new ServiceClient($config->mcpUrl, $transport, $tokenSource));
        $this->tradingControl = new TradingControlClient(new ServiceClient($config->tradingControlUrl, $transport, $tokenSource));
        $this->matchEngine = new MatchEngineClient($config->matchEngineWsUrl);
        $this->keystore = new KeyStoreClient(new ServiceClient($config->keystoreUrl, $transport, $tokenSource));
        $this->hummingbot = new HummingbotClient(new ServiceClient($config->hummingbotUrl, $transport, $tokenSource));
        $this->wise = new WiseClient(new ProviderClient(new ServiceClient($config->wiseUrl, $transport, $tokenSource)));
        $this->airwallex = new AirwallexClient(new ProviderClient(new ServiceClient($config->airwallexUrl, $transport, $tokenSource)));
        $this->planner = new PlannerClient(
            $this->mcp,
            $this->orchestrator,
            $this->analytics,
            $this->tradingControl,
        );
    }

    public static function fromEnv(?callable $transport = null): self
    {
        return new self(SdkConfig::fromEnv(), $transport);
    }
}

File: java/src/main/java/com/twofinance/sdk/SdkClient.java
package com.twofinance.sdk;

import java.net.http.HttpClient;

public final class SdkClient {
    public final SDKConfig config;
    public final DomainClients.AuthClient auth;
    public final DomainClients.NetworkClient network;
    public final DomainClients.AnalyticsClient analytics;
    public final DomainClients.OrchestratorClient orchestrator;
    public final DomainClients.MCPClient mcp;
    public final DomainClients.TradingControlClient tradingControl;
    public final DomainClients.MatchEngineClient matchEngine;
    public final DomainClients.KeyStoreClient keystore;
    public final DomainClients.HummingbotClient hummingbot;
    public final DomainClients.WiseClient wise;
    public final DomainClients.AirwallexClient airwallex;
    public final DomainClients.PlannerClient planner;

    public SdkClient(SDKConfig config) {
        this(config, HttpClient.newHttpClient());
    }

    public SdkClient(SDKConfig config, HttpClient httpClient) {
        this.config = config;
        TokenSource tokenSource = config.tokenSource;
        this.auth = new DomainClients.AuthClient(config.authUrl, httpClient, tokenSource, config);
        this.network = new DomainClients.NetworkClient(config.networkUrl, httpClient, tokenSource);
        this.analytics = new DomainClients.AnalyticsClient(config.analyticsUrl, httpClient, tokenSource);
        this.orchestrator = new DomainClients.OrchestratorClient(config.orchestratorUrl, httpClient, tokenSource);
        this.mcp = new DomainClients.MCPClient(config.mcpUrl, httpClient, tokenSource);
        this.tradingControl = new DomainClients.TradingControlClient(config.tradingControlUrl, httpClient, tokenSource);
        this.matchEngine = new DomainClients.MatchEngineClient(config.matchEngineWsUrl);
        this.keystore = new DomainClients.KeyStoreClient(config.keyStoreUrl, httpClient, tokenSource);
        this.hummingbot = new DomainClients.HummingbotClient(config.hummingbotUrl, httpClient, tokenSource);
        this.wise = new DomainClients.WiseClient(config.wiseUrl, httpClient, tokenSource);
        this.airwallex = new DomainClients.AirwallexClient(config.airwallexUrl, httpClient, tokenSource);
        this.planner = new DomainClients.PlannerClient(this.mcp, this.orchestrator, this.analytics, this.tradingControl);
    }

    public static SdkClient fromEnv() {
        return new SdkClient(SDKConfig.fromEnv());
    }
}
